samplers/aws/xray/remote_sampler.go

```diff
@@ -0,0 +1,219 @@
+type remoteSampler struct {
+	// List of known centralized sampling rules.
+	manifest *manifest
+
+	// xrayClient is used for getting quotas and sampling rules.
+	xrayClient *xrayClient
+
+	// pollerStart, if true represents rule and target pollers are started.
+	pollerStart bool
+
+	// samplingRulesPollingInterval, default is 300 seconds.
+	samplingRulesPollingInterval time.Duration
+
+	// Unique ID used by XRay service to identify this client.
+	clientID string
+
+	// Provides system time.
+	clock clock
+
+	mu sync.RWMutex
+}
+
+// Compile time assertion that remoteSampler implements the Sampler interface.
+var _ sdktrace.Sampler = (*remoteSampler)(nil)
+
+// NewRemoteSampler returns a sampler which decides to sample a given request or not
+// based on the sampling rules set by users on AWS X-Ray console. Sampler also periodically polls
+// sampling rules and sampling targets.
+func NewRemoteSampler(ctx context.Context, opts ...Option) (sdktrace.Sampler, error) {
+	cfg := newConfig(opts...)
+
+	// Generate clientID
+	var r [12]byte
+
+	_, err := crypto.Read(r[:])
+	if err != nil {
+		return nil, fmt.Errorf("unable to generate client ID: %w", err)
+	}
+
+	id := fmt.Sprintf("%02x", r)
+
+	clock := &defaultClock{}
+
+	m := &manifest{
+		rules: []*rule{},
+		index: map[string]*rule{},
+		clock: clock,
+	}
+
+	remoteSampler := &remoteSampler{
+		clock:                        clock,
+		manifest:                     m,
+		clientID:                     id,
+		xrayClient:                   newClient(cfg.endpoint),
+		samplingRulesPollingInterval: cfg.samplingRulesPollingInterval,
+	}
+
+	// starts the rule poller
+	remoteSampler.start(ctx)
+
+	return remoteSampler, nil
+}
+
+func (rs *remoteSampler) ShouldSample(parameters sdktrace.SamplingParameters) sdktrace.SamplingResult {
+	// ToDo: add business logic for remote sampling
+
+	return sdktrace.TraceIDRatioBased(0.05).ShouldSample(parameters)
+}
+
+func (rs *remoteSampler) Description() string {
+	return "remote sampling with AWS X-Ray"
+}
+
+func (rs *remoteSampler) start(ctx context.Context) {
+	if !rs.pollerStart {
+		rs.pollerStart = true
+		rs.startRulePoller(ctx)
+	}
+}
+
+func (rs *remoteSampler) startRulePoller(ctx context.Context) {
+	go func() {
+		// Period = 300s, Jitter = 5s
+		t := newTicker(rs.samplingRulesPollingInterval, 5*time.Second)
+
+		// Periodic manifest refresh
+		for {
+			if err := rs.refreshManifest(ctx); err != nil {
+				globalLogger.Error(err, "Error occurred while refreshing sampling rules")
+			} else {
+				globalLogger.Info("Successfully fetched sampling rules")
+			}
+			select {
+			case _, more := <-t.C():
+				if !more {
+					return
+				}
+				continue
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+func (rs *remoteSampler) refreshManifest(ctx context.Context) (err error) {
+	// Compute 'now' before calling GetSamplingRules to avoid marking manifest as
+	// fresher than it actually is.
+	now := rs.clock.now().Unix()
```

Comment:
Store `time.Time`, don't use `Unix()` anywhere here since we want to use monotonic time